### PR TITLE
Xrpl client wip

### PR DESCRIPTION
**Pull request by bmurphy333**

### CODE REVIEW

File: client/jsonrpc_config.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+	Do(req *http.Request) (*http.Response, error)
+}
+
+type JsonRpcConfig struct {
+	HTTPClient HTTPClient
+	Url        string
+	Headers    map[string][]string
+}
+
+func NewJsonRpcConfigWithHttpClient(url string, c HTTPClient) (*JsonRpcConfig, error) {
```
Comment (JCrawsh) [inline client/jsonrpc_config.go]: Could we convert this to JsonRpcConfigOpts e.g. something like:
```
type JsonRpcClientOpt func(c *JsonRpcClient)

func WithHttpClient(cl HTTPClient) JsonRpcClientOpt {
	return func(c *JsonRpcClient) {
		c.client = cl
	}
}

func NewJsonRpcClient(url string, opts ...JsonRpcClientOpt) *JsonRpcClient {
	cl := &JsonRpcClient{
		URL:    url,
		client: &http.Client{},
	}

	for _, opt := range opts {
		opt(cl)
	}
	return cl
}
```

File: client/jsonrpc_config.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
+package client
+
+import (
+	"net/http"
+	"strings"
+	"time"
+)
+
+type EmptyUrlError struct {
```
Comment (JCrawsh) [inline client/jsonrpc_config.go]: if theres no fields lets convert this to a var
```suggestion
var EmptyUrlError = errors.New("Empty port and IP provided")
```

File: model/client/common/xrpl_response.go [hunk 3]
```diff
@@ -0,0 +1,5 @@
+package common
+
+type XRPLResponse interface {
```
Comment (JCrawsh) [inline model/client/common/xrpl_response.go]: I've put this in the client package as I think it makes more sense to keep it there.

File: client/jsonrpc/utils/rpc_mock_client.go
Comment (JCrawsh) [inline client/jsonrpc/utils/rpc_mock_client.go:1]: Does this need to be a separate package or could it live with the json_client_test code? Could maybe also make use of https://github.com/stretchr/testify#mock-package

File: model/client/common/xrpl_response.go (hunk 3 above)
Comment (bmurphy333) [reply model/client/common/xrpl_response.go]: Moved it :)

File: client/jsonrpc/utils/rpc_mock_client.go
Comment (bmurphy333) [reply client/jsonrpc/utils/rpc_mock_client.go:1]: Have moved it into jsonrpcclient package. Will look at those different tests separately so not to block this PR. Will deffo look at it tho it seems much more simple

File: examples/jsonrpc_client_example.go
Comment (JCrawsh) [inline examples/jsonrpc_client_example.go:1]: Can you put this file on the path `examples/jsonrpc-client/`. I'll make a similar folder for websocket

File: client/client.go [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+type Client interface {
+	SendRequest(reqParams XRPLRequest) (XRPLResponse, error)
+}
+
+type XRPLClient struct {
+	Account Account
+}
+
+type XRPLRequest interface {
```
Comment (JCrawsh) [inline client/client.go:11]: Need to add a validate method to this interface

**Review by JCrawsh — CHANGES_REQUESTED** [review]
A couple of small changes but LGTM!

File: client/account.go [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+	GetAccountChannels(req *account.AccountChannelsRequest) (*account.AccountChannelsResponse, XRPLResponse, error)
+}
+
+type accountImpl struct {
+	Client Client
+}
+
+func (a *accountImpl) GetAccountChannels(req *account.AccountChannelsRequest) (*account.AccountChannelsResponse, XRPLResponse, error) {
+
+	err := req.Validate()
```
Comment (JCrawsh) [inline client/account.go]: Is it worth moving this into the send request function so we don't need to repeat it on every function?

File: examples/jsonrpc_client_example.go
Comment (bmurphy333) [reply examples/jsonrpc_client_example.go:1]: shall do :)

File: client/account.go (hunk 5 above)
Comment (bmurphy333) [reply client/account.go]: yeah makes sense, will move now

Comment (bmurphy333) [reply client/account.go]: Means we will leave all the validation testing to the test files in the model folder

**Review by JCrawsh — APPROVED** [review]
Yeah everythings looks good!

**Pull request merged**